File: test/resources/HelloProphecy/prophecy/pipelines/gold_sales/code/gold_sales/graph/gold_sales_by_zip_date.py
from pyspark.sql import *
from pyspark.sql.functions import *
from pyspark.sql.types import *
from prophecy.utils import *
from prophecy.libs import typed_lit
from gold_sales.config.ConfigStore import *
from gold_sales.udfs.UDFs import *

def gold_sales_by_zip_date(spark: SparkSession, in0: DataFrame):
    from pyspark.sql.utils import AnalysisException

    try:
        desc_table = spark.sql("describe formatted `scottdemo`.`gold_sales_by_zip_date`")
        table_exists = True
    except AnalysisException as e:
        table_exists = False

    if table_exists:
        from delta.tables import DeltaTable, DeltaMergeBuilder
        DeltaTable\
            .forName(spark, "`scottdemo`.`gold_sales_by_zip_date`")\
            .alias("target")\
            .merge(
              in0.alias("source"),
              (
                (col("source.zipcode") == col("target.zipcode"))
                & (col("source.order_date") == col("target.order_date"))
              )
            )\
            .whenMatchedUpdateAll()\
            .whenNotMatchedInsertAll()\
            .execute()
    else:
        in0.write.format("delta").mode("overwrite").saveAsTable("`scottdemo`.`gold_sales_by_zip_date`")

File: test/resources/HelloProphecy/prophecy_scala/pipelines/gold_sales/code/src/main/scala/io/prophecy/pipelines/gold_sales/graph/gold_sales_by_zip_date.scala
package io.prophecy.pipelines.gold_sales.graph

import io.prophecy.libs._
import io.prophecy.pipelines.gold_sales.config.Context
import org.apache.spark._
import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.expressions._
import java.time._

object gold_sales_by_zip_date {

  def apply(context: Context, in: DataFrame): Unit =
    in.write
      .format("delta")
      .mode("overwrite")
      .saveAsTable("`scottdemoscala`.`gold_sales_by_zip_date`")

}
